contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,244 @@
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import "../lib/PreciseUnitMath.sol";
+import "../../external/contracts/UniswapV2Library.sol";
+import "../../external/contracts/SushiswapV2Library.sol";
+import "../interfaces/ISetToken.sol";
+import "../interfaces/IBasicIssuanceModule.sol";
+import "../interfaces/IWETH.sol";
+
+/**
+ * @title ExchangeIssuance
+ * @author Noah Citron
+ *
+ * Contract for minting and redeeming any Set token using
+ * ETH as the paying/receiving currency. All swaps are done using the best price
+ * found on Uniswap or Sushiswap.
+ *
+ */
+contract ExchangeIssuance {
+
+    using SafeMath for uint256;
+    
+    /* ============ State Variables ============ */
+
+    IUniswapV2Router02 private uniRouter;
+    address private uniFactory;
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(address indexed recipient, address indexed index, address indexed inputToken, uint256 amount);
+    event ExchangeRedeem(address indexed recipient, address indexed index, address indexed outputToken, uint256 amount);
+
+    /* ============ Constructor ============ */
+
+    constructor(address _uniFactory, address _uniRouter, address _sushiFactory, address _sushiRouter, address _basicIssuanceModule) public {
+        uniFactory = _uniFactory;
+        uniRouter = IUniswapV2Router02(_uniRouter);
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = IUniswapV2Router02(_sushiRouter);
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = IBasicIssuanceModule(_basicIssuanceModule);
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming an index. This function must only be called before the first time
+     * this smart contract is used on any particular index, or when a new token is added
+     * to the index.
+     *
+     * @param _index    Address of the index being initialized
+     */
+    function initApprovals(address _index) external {
+        ISetToken.Position[] memory positions = ISetToken(_index).getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(basicIssuanceModule), PreciseUnitMath.maxUint256());
+        }
+    }
+
+    /**
```

Comment:
Since this function calls potentially untrusted smart contracts, let's add a ReEntrancy guard on `exchangeIssue` and `exchangeRedeem` . You should be able to find it in the OpenZeppelin library